pkg/executor/sortexec/topn.go

```diff
@@ -17,147 +17,192 @@ package sortexec
+	//
+	// This variable is only used for test.
+	isSpillTriggeredInStage1ForTest bool
+	isSpillTriggeredInStage2ForTest bool
 }
 
-// Less implement heap.Interface, but since we mantains a max heap,
-// this function returns true if row i is greater than row j.
-func (h *topNChunkHeap) Less(i, j int) bool {
-	rowI := h.rowChunks.GetRow(h.rowPtrs[i])
-	rowJ := h.rowChunks.GetRow(h.rowPtrs[j])
-	return h.greaterRow(rowI, rowJ)
-}
+// Open implements the Executor Open interface.
+func (e *TopNExec) Open(ctx context.Context) error {
+	concurrency := e.Ctx().GetSessionVars().Concurrency.ExecutorConcurrency
+	e.memTracker = memory.NewTracker(e.ID(), -1)
+	e.memTracker.AttachTo(e.Ctx().GetSessionVars().StmtCtx.MemTracker)
 
-func (h *topNChunkHeap) greaterRow(rowI, rowJ chunk.Row) bool {
-	for i, colIdx := range h.keyColumns {
-		cmpFunc := h.keyCmpFuncs[i]
-		cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
-		if h.ByItems[i].Desc {
-			cmp = -cmp
-		}
-		if cmp > 0 {
-			return true
-		} else if cmp < 0 {
-			return false
+	e.fetched = &atomic.Bool{}
+	e.fetched.Store(false)
+	e.chkHeap = &topNChunkHeap{memTracker: e.memTracker}
+	e.chkHeap.idx = 0
+
+	e.finishCh = make(chan struct{}, 1)
+	e.resultChannel = make(chan rowWithError, e.MaxChunkSize())
+	e.chunkChannel = make(chan *chunk.Chunk, concurrency)
+	e.inMemoryThenSpillFlag = false
+	e.isSpillTriggeredInStage1ForTest = false
+	e.isSpillTriggeredInStage2ForTest = false
+
+	if variable.EnableTmpStorageOnOOM.Load() {
+		e.diskTracker = disk.NewTracker(e.ID(), -1)
+		e.diskTracker.AttachTo(e.Ctx().GetSessionVars().StmtCtx.DiskTracker)
+		e.fetcherAndWorkerSyncer = &sync.WaitGroup{}
+
+		workers := make([]*topNWorker, concurrency)
+		for i := range workers {
+			chkHeap := &topNChunkHeap{}
+			// Offset of heap in worker should be 0, as we need to spill all data
+			chkHeap.init(e, e.memTracker, e.Limit.Offset+e.Limit.Count, 0, e.greaterRow, e.RetFieldTypes())
+			workers[i] = newTopNWorker(i, e.chunkChannel, e.fetcherAndWorkerSyncer, e.resultChannel, e.finishCh, e, chkHeap, e.memTracker)
 		}
-	}
-	return false
-}
 
-func (h *topNChunkHeap) Len() int {
-	return len(h.rowPtrs)
-}
+		e.spillHelper = newTopNSpillerHelper(
+			e,
+			e.finishCh,
+			e.resultChannel,
+			e.memTracker,
+			e.diskTracker,
+			exec.RetTypes(e),
+			workers,
+			concurrency,
+		)
+		e.spillAction = &topNSpillAction{spillHelper: e.spillHelper}
+		e.Ctx().GetSessionVars().MemTracker.FallbackOldAndSetNewAction(e.spillAction)
+	}
 
-func (*topNChunkHeap) Push(any) {
-	// Should never be called.
+	return exec.Open(ctx, e.Children(0))
 }
 
-func (h *topNChunkHeap) Pop() any {
-	h.rowPtrs = h.rowPtrs[:len(h.rowPtrs)-1]
-	// We don't need the popped value, return nil to avoid memory allocation.
-	return nil
-}
+// Close implements the Executor Close interface.
+func (e *TopNExec) Close() error {
+	close(e.finishCh)
+	if e.fetched.CompareAndSwap(false, true) {
+		close(e.resultChannel)
+		return nil
```

Comment:
> And then `exec.Close(e.Children(0))` will be skipped? why not `close(e.resultChannel)` directly?

It's a mistake, I will add `exec.Close(e.Children(0))`. 

`e.resultChannel` will be closed in `executeTopN` and it will cause the close of closed channel if we `close(e.resultChannel)` directly.